probe/appclient/multi_client.go

```diff
@@ -166,13 +166,21 @@ func (c *multiClient) Stop() {
 // reader, and recreate new readers for each publisher. Note that it will
 // publish to one endpoint for each unique ID. Failed publishes don't count.
 func (c *multiClient) Publish(r io.Reader) error {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if len(c.clients) <= 1 { // optimisation
+		for _, c := range c.clients {
+			return c.Publish(r)
+		}
+		return nil
+	}
+
 	buf, err := ioutil.ReadAll(r)
 	if err != nil {
 		return err
 	}
 
-	c.mtx.Lock()
-	defer c.mtx.Unlock()
```

Comment:
Minor downside here is that the lock now covers the `ioutil.ReadAll(r)`. That's in the >1 case only though, so not the common path, and eliminating it would require unlocking and relocking which may well be more expensive.